### PR TITLE
gadgets: Automatically add annotations based on the field types

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
Gadgets use a lot of well-known fields like pid, tid, uid. etc. This commit adds logic in Inspektor Gadget to automatically add annotations for those types to avoid having to repeat the same thing in the metadata file for all gadgets. It is still possible for gadget developers to override the defaults.

Part of #1567

### TODO
- [ ]  Update hello-world guide (will be done in (#3506)

### CODE REVIEW

**Comment by flyth** [comment]
This looks good! - I think it should be documented somewhere [here](https://www.inspektor-gadget.io/docs/latest/gadget-devel/gadget-ebpf-api#enriched-types) that those types will automatically get annotations.

I know this PR isn't the source of it, but I'd really prefer if we keep IG specific things outside of the `datasource` package.

**Comment by mauriciovasquezbernal** [comment]
> This looks good! - I think it should be documented somewhere [here](https://www.inspektor-gadget.io/docs/latest/gadget-devel/gadget-ebpf-api#enriched-types) that those types will automatically get annotations.
> 

I added a small sentence. It's still missing to document the types and update the hello-world gadget, but I prefer to do that on #1567 as this PR in an intermediate step towards it.

> I know this PR isn't the source of it, but I'd really prefer if we keep IG specific things outside of the `datasource` package.

I tried to move things around a bit, I'm not totally sure about the result. Please let me know what you think, otherwise I'll revert it and fix it later on.

File: pkg/metadata/v1/annotations.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadatav1
+
+// This file contains all
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/metadata/v1/annotations.go]: Removed this.

File: pkg/datasource/data.go [hunk 2]
```diff
@@ -472,11 +466,23 @@ func (ds *dataSource) applyFieldConfig(newFields ...*field) {
 
 	// first pass applying configuration from metadata
 	for _, field := range newFields {
+		// apply template based on type of the field
+		for _, tag := range field.Tags {
+			suffix, ok := strings.CutPrefix(tag, "type:")
+			if !ok {
+				continue
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/datasource/data.go:472]: This code doesn't exist on the last iteration.

File: include/gadget/types.h [hunk 3]
```diff
@@ -59,6 +59,17 @@ typedef __u64 gadget_syscall;
...
+typedef __u32 gadget_pid;
+typedef __u32 gadget_ppid;
+typedef __u32 gadget_tid;
+
+#ifndef TASK_COMM_LEN
+#define TASK_COMM_LEN 16
+#endif
+
+typedef char gadget_comm[TASK_COMM_LEN];
+typedef char gadget_pcomm[TASK_COMM_LEN];
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply include/gadget/types.h:71]: What are your specific concerns about this one? We can use `typedef char gadget_comm` and then let the user specify the size on the event definition. But I want to know more why this approach could lead to troubles.

[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply include/gadget/types.h]: That makes sense. I updated them to be `typedef char gadget_{p}comm`

**Review by burak-ok — APPROVED** [review]
From my side the code looks clean and removes a lot of duplication as you described.

Tested locally a little bit and works as advertised. Thanks

File: pkg/datasource/templates.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	annotationMutex.Lock()
+	defer annotationMutex.Unlock()
+
+	handled := false
+	for _, cb := range annotationTemplateCallbacks {
+		if ok := cb(name, annotations); ok {
+			handled = true
+		}
+	}
+	return handled
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/datasource/templates.go:44]: It's not equivalent. We want to process **all** callbacks and return true if any of the callbacks handled it. Your suggestion will only process the first callback.

**Comment by mauriciovasquezbernal** [comment]
Thanks folks the all the suggestions and feedback. I really enjoyed this one.

**Comment by MeherRushi** [comment]
Hey, I think we need to add the gadget_process respective structs to the types.go of the utils packages right for testing purposes, right? If so, I will make the necessary changes and make a PR.

**Comment by mauriciovasquezbernal** [comment]
@MeherRushi makes sense to me!

**Comment by MeherRushi** [comment]
@mauriciovasquezbernal @eiffel-fl 

Hey so I was actually going through where the structs were present and I realized that the `Process` struct was actually already defined in `types.go` of the `operator/ebpf/types` package and can be used from there

I am not sure if writing the same in the `testing/utils/types.go` will be redundant or not.

As I just realized that L3endpoint has 2 definitions, one in the `types.go` of `testing/utils` package and `types.go` of the `operators/ebpf/types` package.
I am a little confused as to which is to be used when exactly ? Is the `types.go` in of the places redundant ?

In the `types.go of operators/ebpf/types package`
```C
// L3Endpoint is the Golang representation of struct gadget_l3endpoint_t
type L3Endpoint struct {
	V6      [16]byte
	Version uint8
	_       [3]byte
}
```

in the `types.go of testing/utils package`
```C
type L3Endpoint struct {
	Addr    string `json:"addr"`
	Version uint8  `json:"version"`
}
```

I think it is redundant to have it both places. If so, I'll not change anything, else I'll add the structs to test/utils package types.go file

**Comment by mauriciovasquezbernal** [comment]
> I am not sure if writing the same in the testing/utils/types.go will be redundant or not.

I think that would be redundant. 

> I am a little confused as to which is to be used when exactly ? Is the types.go in of the places redundant ?

Those are a little bit different. The one defined in pkg/operators/ebpf/types/types.go only has the fields provided by the ebpf program (low level information), while the one defined in pkg/testing/utils/types.go contains enriched fields added by the formatter operator. I don't think there is an easy way to combine them. 

The only idea that I have is to use alias names in pkg/testing/utils/types.go, so users only need to import this package in the testing. Something like:

```diff
diff --git a/pkg/testing/utils/types.go b/pkg/testing/utils/types.go
index e3a0fba22..cb28b2be9 100644
--- a/pkg/testing/utils/types.go
+++ b/pkg/testing/utils/types.go
@@ -14,6 +14,8 @@

 package utils

+import ebpftypes "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ebpf/types"
+
 // Define some types that are only used for testing purposes

 type K8s struct {
@@ -35,3 +37,5 @@ type L3Endpoint struct {
        Addr    string `json:"addr"`
        Version uint8  `json:"version"`
 }
+
+type Process ebpftypes.Process
```

What do you think?

**Comment by MeherRushi** [comment]
Right, that makes sense. I can do that.

I think a few gadgets have already imported `ebpftypes "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/ebpf/types"` in their test files.

I can maybe make amends in all the places if that's okay?

**Comment by MeherRushi** [comment]
@mauriciovasquezbernal @eiffel-fl 

I did make a PR for the same. #3764 

However, instead of having a new type using `type Process ebpftypes.Process` , I have created a alias using `type Process = ebpftypes.Process` in the utils package to make the usage easier and avoid type conversion in a few functions that have the argument and return type as `ebpftypes.Process`

**Pull request merged**